README.md
berry4all
=========

Software that allows tethering using a blackberry on Linux / Mac

At this point (2013) it's probably fairly obsolete however there are still eople using it.

It was quite popular in the past (2005-2012).

More info here: http://www.berry4all.com/


ChangeLog.txt
BBTether
Thibaut Colar
tcolar AT colar d.o.t NET
0.3p: 3/25/11:
	New config files
	OSX fixes
0.3m: 11/29/10:
	Config files fixes
	Many new ISP config files
0.3l: 04/25/10:
	Config files fixes
	And new config files for other providers
0.3k: 11/5/09:
	New orange config (Suiss)
	New USCellular config
	Another password routine fix
0.3j: 9/21/09:
	Password routine fix (Again)

0.3h: 9/10/09:
	Fixed bug in password routine
	Added some more user provided ppp scripts

0.3g: 4/29/2009:
	Remove uses of BaseException.message (deprecated in python 2.6)

0.3f: 4/17/2009:
	Added GUI option to force modem data mode only, might help on some devices.

0.3e: 4/8/2009:
	Fixes to some ppp scripts

0.3d: 3/29/2009:
	Fix to password seed verification code.
	Other small bug fixes (debug calls)

0.3c: 3/26/2009:
	If stable, will not changed for a little while (stable version)
	Fixed bug in preferences handling (could cause endpoint scanning to fail and thus not usable)
	Removed obsolete wx entries (replaced by wx.)
	Addec Max down/UP transfer speed to GUI.

0.3b: 3/24/2009:
	More GUI work, now displays: IP, peer IP, DNS 1 & 2 IP
	Also displays speed stats (current, average, total) and connection time.

0.3a: 3/24/2009:
	Lots of GUI work, gui should now be fully working.
	Other fixes and cleanup.
	berry4all.sh script to start the GUI

0.3beta6: 3/21/2009:
	Fixed a bad bug that would cause a diconect after ~100k sent

0.3beta5: 3/20/2009:
	Fixes to password support
	MUltiple small bug fixes
	GUI now usable (asks for pppd conf & password)

0.3beta4: 3/20/2009:
	Added support for detaching binded usb kernel, should
	take care of those can't claim interface issues. (Linux only)
	Renamed gui to Berry4All

0.3beta2: 3/20/2009:
	More failure checks and debug infos
	.pyc clenup shell script (cleanup.sh)

0.3beta2: 3/20/2009:
	Fixed to save endpoints routine

0.3beta1: 3/19/2009:
	Started GUI work - not fully complete (bbgui)
	Fixed bbtether broken endpoint save/load in 0.2l
	Fixed command line shutdown
	Fixed error reported by pylint

0.2l: 3/16/2009:
	Save/Load endpoints data, so we only scan once, this fixes 2 issues:
	- Makes subsequent use of bbtether faster
	- Skipping the scan works better on some devices (esp. bold)
	Works on Bold (OS X)

0.2k: 3/15/2009:
	More OS X work, check for poketmac (conflict) etc...

0.2j: 3/14/2009:
	Fixed code so that it can work with libusb binary
	Had to ignore errors on OSX thrown on usb timeouts (empty errors - no numbers)
	OSX: create pap-secrets as needed.

0.2i: 3/14/2009:
	Fixes but in modem readline code, was causing failure of ppp
	negocation (on mac at least)
	Added "dump" options to pppd when in verbose mode
	Works on OS X once again.

0.2h: 3/13/2009:
	Fixed bug in password code (password protected devices)
	Working on Curve ! (8320)
	
0.2g: 3/12/2009:
	Made charging mode switch (more power) optional, as this does not work/not needed on storm.
	Replaced SHA1 library(deprecated) by the new Hashlib
	Only reset device if needed (no reply to modem start - Pearl)
	Try modem init cycle multiple times (incl. reset) until good to go.
	Added "Data Mode" flag option (-m)
	Working on Storm! (9500)

0.2f: 3/10/2009
	Added password protected device support (tested on Pearl)
	Added (to be tested) support for ppp that sends end of line as 0xD rather than 0xD,0xA

0.2e: 3/9/2009
	Fixed pp synchronization issues
	Needed to read line one at a time
	Should prevent random ppp chat failures
	OS X broke again :-(

0.2d: 3/8/2009
	Fixed some ppp scripts again (simpified)
	First version to work on Mac OSX (need to document)

0.2c: 3/8/2009
	Rewrote ppp scripts to be like traced on windows
	Fixed double ppp frames issue (0x7E needed to be doubled)
	Fixed read buffer sizes and timeouts values
	Now seem to work good and not disconnect/hang anymore (tested >30mn)
	Got speedtest.net to work fully (down and up), upload was failing before
	Tested working on rogers(CA) by Max Taranukha

0.2b: 3/4/2009
	Should be able to do multiple sessions without need to reboot blackberry
	Should be able to have a session running for a long time (no more disconnect after few minutes)

0.2a: 3/4/2009
	Many Changes, to make it work with OS 4.5 (did not work before)
	Should be faster (better USB lattency)
	More ppp scripts (& t-mobile options)
	Better Error Catching
	Cleaner shutdowns

0.1p: 2/19/2009
	Rogers (canada) ppp scripts

0.1n: 2/15/2009
	Prevent program failure if setting power mode / data mode fails (9500)
	Added UDEV rules to be able to run as regular user.

0.1m: 2/14/2009
	Does not crash after usb read failure, keep going
	Now works on 8130 (pearl 2)
	If finds data/modem pairs, does not scan other interfaces

0.1l: 2/13/2009
	-i option to force interface - doesn't check other ones
	Fixed some more endpoints detection
	Replace root requirement with warning

0.1k: 2/13/2009
	Improved modem detection (some device reply to hello!)
	Removed full usb modules listing

0.1j: 2/13/2009
	Added more berry_chrage checks
	Prevent complete failure if one interface is not avilable (v2)
	cleanups

0.1i: 2/13/2009
	Added some more osx support code
	Prevent complete failure if one interface is not avilable (busy etc...)

0.1h: 2/12/2009
	Fixed various typos

0.1g: 2/12/2009
	Fix so that it terminates correctly (use to hang)
	Fix so that it terminates correctly (use to hang)

0.1f: 2/12/2009
	Added ppp scripts for sprint & att
	Added link to site in output

0.1e: 2/12/2009
	Introduced this file (ChangeLog)
	Added tentative porta (Ecuador) ppp conf
	Check running as root
	Try to warn / unload berry_charge module

0.1d: 2/11/2009
	Removed changdefaultroute from ppp option since not always avail.
	Updated doc to match that
	Added some OSX code (not yet used)
	Added usb/protocol references

0.1c: 2/10/2009
	Refactoring (not all in one file)
	Added bulid script
	Added PyLint checks and fixed what was found
	Improved pppd output grabbing, for debugging.

0.1b: 2/10/2009
	Fix chat script
	Fix bug causing disconnects (0x7E frames)

bbtether.py
#!/usr/bin/python

'''
See: http://wiki.colar.net/bbtether for more infos.

Script to use a USB BlackBerry device as a modem (Tethering), and enable charging.
On My Pearl (Edge) in Seattle, I get speeds of 8K/s-29K/s, avering about 14K/s.

This script requires python, pppd, libusb and the python usb module installed:
	You probably already have python, pppd and libusb installed
	Ex: sudo apt-get install python libusb pppd python-pyusb
		or yum install python libusb pppd pyusb

-----------------------------------		
Thibaut Colar - 2009+
tcolar AT colar DOT net
http://wiki.colar.net/bbtether
Released Under GPL2, COMES WITH ABSOLUTELY NO WARRANTIES OF ANY KIND, USE AT YOUR OWN RISK.
If you make fixes or find issues please EMAIL: tcolar AT colar Dot NET
'''
 
import usb
import pty
import os
import time
import threading
import array
import string
import subprocess
from optparse import OptionParser, OptionGroup

VERSION="0.1"

# you MIGHT need to enter the full path to pppd here
PPPD_COMMAND="pppd"

VENDOR_RIM=0x0fca
PRODUCT_DATA=0x0001   #(serial data)
PRODUCT_NEW_DUAL=0x0004   #(mass storage & data)
PRODUCT_NEW_8120=0x8004   #(Pearl 8120)
PRODUCT_NEW_MASS_ONLY=0x0006   #(mass storage only)

BERRY_CONFIG=1

TIMEOUT=1000
BUF_SIZE=25000
MIN_PASSWD_TRIES=2
#notify every aprox ~ 100K bytes transferred
NOTIFY_EVERY=100000

COMMAND_PIN = [0x00,0x00,0x0c,0x00,0x05,0xff,0x00,0x00,0x00,0x00,0x04,0x00] 
COMMAND_DESC= [0x00,0x00,0x0c,0x00,0x05,0xff,0x00,0x00,0x00,0x00,0x02,0x00]
COMMAND_HELLO = [0x00, 0x00, 0x10, 0x00, 0x01, 0xff, 0x00, 0x00,0xa8, 0x18, 0xda, 0x8d, 0x6c, 0x02, 0x00, 0x00]
MODEM_CONNECT= [0xd, 0xa, 0x43, 0x4f, 0x4e, 0x4e, 0x45, 0x43, 0x54, 0xd, 0xa]
MODEM_START = [0x01, 0x00, 0x00, 0x00, 0x78, 0x56, 0x34, 0x12]
MODEM_STOP = [0x01, 0x00 ,0x00, 0x00 ,0x00, 0x00, 0x00, 0x00, 0x78, 0x56, 0x34, 0x12]
# Packets ending by this are RIM control packets (! data)
RIM_PACKET_TAIL=[0x78, 0x56, 0x34, 0x12]
	
''' Main Class '''
class BBTether:
	verbose=False;
	ppp=None

	def find_berry(self, userdev, userbus, verbose=True):
		if verbose:
			print "Looking for USB devices:"
		berry=None
		if userdev and userbus:
			if verbose :
				print "Will use user provided bus/device: ",options.bus,"/",options.device
			for bus in usb.busses():
				if string.atoi(bus.dirname) == string.atoi(userbus):
					for dev in bus.devices:
						if string.atoi(dev.filename) == string.atoi(userdev):
							berry=dev				
		else:
			for bus in usb.busses():
				for dev in bus.devices:
					if(verbose):
						print "	Bus %s Device %s: ID %04x:%04x" % (bus.dirname,dev.filename,dev.idVendor,dev.idProduct)
					if(dev.idVendor==VENDOR_RIM):
						berry=dev
		return berry


	def set_bb_power(self,handle):
		print "\nIncreasing USB power - for charging"
		buffer= [0,0]
		handle.controlMsg(0xc0, 0xa5, buffer, 0 , 1)
		buffer = []
		handle.controlMsg(0x40, 0xA2, buffer, 0 , 1)
		handle.reset()
		print "Switching Device to data only mode"
		buffer= [0,0]
		handle.controlMsg(0xc0, 0xa9, buffer, 0 , 1)

	def readlong(self,data,index):
		return (data[index+3]<<24)+(data[index+2]<<16)+(data[index+1]<<8)+data[index]

	def readstring(self,data,index):
		s="";
		while(data[index]!=0):
			s+=chr(data[index])
			index+=1
		return s;

	def get_pin(self, handle, writept, readpt):
		pin=0x0;
		self.usb_write(handle,writept, COMMAND_PIN)
		data=self.usb_read(handle,readpt);
		if data[4] == 0x6 and data[10] == 4:
			pin=self.readlong(data,16);
		return pin

	def get_description(self, handle, writept, readpt):
		desc=""
		self.usb_write(handle,writept, COMMAND_DESC)
		data=self.usb_read(handle,readpt);
		if data[4] == 0x6 and data[10] == 2:
			desc=self.readstring(data,28)
		return desc

	def parse_cmd(self):
		usage = usage = "usage: %prog [options] [pppscript]\n	If [pppscript] is there (ppp script in conf/ folder, ex: tmobile) then will start modem and connect using that ppp script. Otherwise just opens the modem and you will have to start pppd manually."
		parser = OptionParser(usage)
		parser.add_option("-l", "--list", action="store_true", dest="listonly",help="Only detect and list Device, do nothing more")
		parser.add_option("-p", "--pppd", dest="pppd",help="Path To pppd binary (default: /usr/sbin/pppd)")
		parser.add_option("-v", "--verbose", action="store_true", dest="verbose",help="Verbose: Show I/O data and other infos")
		parser.add_option("-c", "--charge", action="store_true", dest="chargeonly",help="Put the device in Charging mode and does NOT start modem.")		
		group = OptionGroup(parser, "Advanced Options","Don't use unless you know what you are doing.")
		group.add_option("-w", "--drp", dest="drp", help="Force Data read endpoint - Hex(ex -w 0x84)")
		group.add_option("-x", "--dwp", dest="dwp", help="Force Data write endpoint - Hex (ex -x 0x6)")
		group.add_option("-y", "--mrp", dest="mrp", help="Force Modem read endpoint - Hex (ex -y 0x85)")
		group.add_option("-z", "--mwp", dest="mwp", help="Force Modem write endpoint - Hex (ex -z 0x8)")
		group.add_option("-d", "--device", dest="device", help="Force to use a specific device ID (use together with -b)")
		group.add_option("-b", "--bus", dest="bus", help="Force to use a specific bus ID(use together with -d)")
		parser.add_option_group(group)
		return parser.parse_args()

	def usb_write(self,handle,pt,bytes,timeout=TIMEOUT,msg="\t-> "):
		if(self.verbose):	
			debug_bytes(bytes,msg)
		try:
			handle.bulkWrite(pt, bytes, TIMEOUT)
		except usb.USBError, error:
			if error.message != "No error":
				raise
				
	def usb_read(self,handle,pt,size=BUF_SIZE,timeout=TIMEOUT,msg="\t<- "):
		bytes=handle.bulkRead(pt, size, TIMEOUT)
		if(self.verbose):	
			debug_bytes(bytes,msg)
		return bytes 

	def __init__(self):
		print "--------------------------------"
		print "BBTether ",VERSION
		print "Thibaut Colar - 2009"
		print "http://wiki.colar.net/bbtether"
		print "Use '-h' flag for more informations : 'python bbtether.py -h'."
		print "--------------------------------\n"
		
		(options,args)=self.parse_cmd()
		self.ppp=None
		if len(args) > 0:
			self.ppp=args[0]
		
		if(options.verbose):
			self.verbose=True
		
		berry=None
		handle=None
		readpt=-1;
		writept=-1;
		interface=-1;
		modem_readpt=-1
		modem_writept=-1

		berry=self.find_berry(options.device,options.bus)				

		if(berry != None):
			# open the connection
			handle=berry.open()
			# set power & reset
			self.set_bb_power(handle)
			
			if options.chargeonly:
				print "Charge only requested, stopping now."
				os._exit(0)
			
			# reopen
			print ("Waiting few seconds, for mode to change")
			time.sleep(1.5)
			
			berry=self.find_berry(options.device,options.bus,False)
			handle=berry.open()
				
			# List device Infos for information and find USB endpair	
			config=berry.configurations[0];
			type=""
			if(berry.idProduct == PRODUCT_DATA):
				type="Data Mode"
			if(berry.idProduct == PRODUCT_NEW_DUAL):
				type="Dual Mode"
			if(berry.idProduct == PRODUCT_NEW_8120):
				type="8120"
			if(berry.idProduct == PRODUCT_NEW_MASS_ONLY):
				type="Storage Mode"
			
			print "\nFound RIM device (",type,")" 
			print "	Manufacturer:",handle.getString(berry.iManufacturer,100)
			print "	Product:",handle.getString(berry.iProduct,100)
			#print "	Serial:",handle.getString(berry.iSerialNumber,100)
			print "	Device:", berry.filename
			print "	VendorId: %04x" % berry.idVendor
			print "	ProductId: %04x" % berry.idProduct
			print "	Version:",berry.deviceVersion
			print "	Class:",berry.deviceClass," ",berry.deviceSubClass
			print "	Protocol:",berry.deviceProtocol
			print "	Max packet size:",berry.maxPacketSize
			print "	Self Powered:", config.selfPowered
			print "	Max Power:", config.maxPower
			for inter in config.interfaces:
				print "	Interface:",inter[0].interfaceNumber
				print "		Interface class:",inter[0].interfaceClass,"/",inter[0].interfaceSubClass
				print "		Interface protocol:",inter[0].interfaceProtocol
				for att in inter:
					i=0
					# check endpoint pairs
					while i < len(att.endpoints):
						good=False
						red=att.endpoints[i].address
						writ=att.endpoints[i+1].address
						i+=2
						print "		EndPoint Pair:",hex(red),"/",hex(writ)
						try:						
							self.usb_write(handle,writ,COMMAND_HELLO)
							try:
								self.usb_read(handle, red)
								good=True
								if readpt == -1 :
									# Use first valid data point found
									interface=inter[0].interfaceNumber
									readpt=red
									writept=writ
							except usb.USBError:
								print "			Not Data Pair (Read failed)"
						except usb.USBError:
							print "			Not Data Pair (Write failed)"
						
						if good:
							print "			Found Data pair:",hex(red),"/",hex(writ);
						else:
							if readpt != -1 and modem_readpt == -1:
								# use pair after data pair as Modem pair
								modem_readpt=red
								modem_writept=writ

			if options.drp:
				readpt=int(options.drp, 16)
			if options.dwp:
				writept=int(options.dwp, 16)
			if options.mrp:
				modem_readpt=int(options.mrp, 16)
			if options.mwp:
				modem_writept=int(options.mwp, 16)				

			if options.listonly :
				print "Listing only requested, stopping here."
				os._exit(0)

			if readpt==-1:
				print "\nNo good Data Endpoint pair, bailing out !";
			else:
				print "\nUsing Data Endpoint Pair:",hex(readpt),"/",hex(writept);				
				print "Using first pair after Data pair as Modem pair: ",hex(red),"/",hex(writ),"\n"
				
				handle.claimInterface(interface);
				
				pin=self.get_pin(handle,writept,readpt)
				print "Pin: ",hex(pin)
				
				desc=self.get_description(handle,writept,readpt)
				print "Description: ",desc

				# Modem use does not require to be in desktop mode, so don't do it.
				#self.enable_desktop_mode(handle,writept,readpt)
				modem=BBModem(self, handle, modem_readpt, modem_writept)
				
				# This will run forever (until ^C)
				pppdCommand="/usr/sbin/pppd";
				if options.pppd:
					pppdCommand=options.pppd
				modem.start(pppdCommand)				
			
		else:
			print "\nNo RIM device found"


class BBModem:		
	'''BlackBerry Modem class'''
	handle=None
	readpt=None
	writept=None
	red=0
	writ=0
	lastcount=0
	
	def __init__(self, prt, hnd, read, write):
		self.handle=hnd
		self.readpt=read
		self.writept=write
		self.parent=prt

	def write(self, data, timeout=TIMEOUT):
		self.parent.usb_write(self.handle,self.writept,data,timeout,"\tModem -> ")
		self.writ+=len(data)
		if(self.red+self.writ>self.lastcount+NOTIFY_EVERY):
			print "GPRS Infos: Received Bytes:",self.red,"	Sent Bytes:",+self.writ
			self.lastcount=self.red+self.writ
			
	def try_read(self, size=BUF_SIZE,timeout=TIMEOUT):
		data=[]
		try:
			data=self.read(size,timeout)
		except usb.USBError, error:
			if error.message != "No error":
				raise
		return data
		
	def read(self, size=BUF_SIZE,timeout=TIMEOUT):
		data=self.parent.usb_read(self.handle,self.readpt,size,timeout,"\tModem <- ")
		self.red+=len(data)
		if(self.red+self.writ>self.lastcount+NOTIFY_EVERY):
			print "GPRS Infos: Received Bytes:",self.red,"	Sent Bytes:",+self.writ
			self.lastcount=self.red+self.writ
		return data

	def init(self):
		'''Initialize the modem and start the RIM session'''
		session_key=[0, 0, 0, 0, 0, 0, 0, 0, 0]
		# clear endpoints
		self.handle.clearHalt(self.readpt)
		self.handle.clearHalt(self.writept)
		# reset modem
		self.write(MODEM_STOP)
		#might or not reply, so use try_read
		self.try_read()
		self.write(MODEM_START)
		answer=self.read()
		# check for password request (newer devices)
		if len(answer)>0 and answer[0]==0x2 and end_with_tuple(answer,RIM_PACKET_TAIL):
			triesLeft=answer[8]
			seed=answer[4:8]
			print "Got password Request from Device (",triesLeft," tries left)"
			triesLeft=answer[8]
			if triesLeft <= MIN_PASSWD_TRIES:
				print "The device has only "+answer[8]+" password tries left, we don't want to risk it! Reboot/unplug the device to reset tries.";
				raise Exception
			#TODO
			print "Password protected Device not supported yet !"
			raise Exception		
		else:
			print "No password requested."	
		# Send session key
		# At least on my Pearl if I don't send this now, the device will reboot itself during heavy data transfer later (odd)
		session_packet=[0, 0, 0, 0, 0, 0, 0, 0, 0x3, 0, 0, 0, 0, 0xC2, 1]+ session_key + RIM_PACKET_TAIL
		self.write(session_packet)
		self.read()
		
	def start(self, pppdCommand):
		'''Start the modem and keep going until ^C'''
		#open modem PTY
		(master,slave)=pty.openpty()
		print "\nModem pty: ",os.ttyname(slave)
				
		print "Initializing Modem"
		try:
			self.init()
		except:
			# If init fails, cleanup and quit
			print "Modem initialization Failed !"
			os.close(master)
			os.close(slave)
			raise

		# Start the USB Modem read thread
		bbThread=BBModemThread(self.handle,self,self.readpt,self.writept,master)
		bbThread.start()
		
		print "Modem Started"
		
		if(not self.parent.ppp):
			print "No ppp requested, you can now start pppd manually."
		else:
			#TODO: start ppp in thread/process
			print "Will try to start pppd now, (",pppdCommand,") with config: ",self.parent.ppp
			time.sleep(.5)
			command=[pppdCommand,os.ttyname(slave),"file","conf/"+self.parent.ppp,"nodetach"]
			if self.parent.verbose:
				command=+"debug"
			pid = subprocess.Popen(command).pid
			# do terintaing this myself, since it should terminate by itself (properly) when bbtether is stopped.
		
		print "Modem Ready at ",os.ttyname(slave)," Use ^C to terminate"
		
		try:
			# Read from PTY and write to USB modem until ^C
			while(True):
				data=os.read(master, BUF_SIZE)
				# transform string from PTY into array of signed bytes
				bytes=array.array("B",data)
				if(len(bytes)>0):
					# start grps fix (needs 0x7E around each Frame)
					newbytes=[]
					last=0
					last2=0
					started=False
					for b in bytes:
						if started and last2 != 0x7e and last == 0x7e and b != 0x7e:
							debug("GPRS fix: Added 0x7E to Frame.")
							newbytes.append(0x7e);
						if not started and b != 0x7e and last == 0x7e:
							started=True
						last2=last
						last=b
						newbytes.append(b)					
					# end gprs fix
					self.write(newbytes)
				
		except KeyboardInterrupt:
			print "\nShutting down on ^C"
			
		# Shutting down "gracefully"
		bbThread.stop()
		os.close(master)
		os.close(slave)
					
class BBModemThread( threading.Thread ):	
	'''Thread that reads Modem data from BlackBerry USB'''

	def __init__(self, hnd, mdm, read, write, master):
		threading.Thread.__init__(self)
		self.output=master
		self.handle=hnd
		self.writept=write
		self.readpt=read
		self.modem=mdm
	
	def stop(self):
		self.done=True

	def run (self):
		self.done=False
		print "Starting Modem thread"
		isFirstConnect=False;
		while( not self.done):
			try:
				try:
					# Read from USB modem and write to PTY
					bytes=self.modem.read()
					if(len(bytes)>0):
						if end_with_tuple(bytes,RIM_PACKET_TAIL):
							# Those are RIM control packet, not data. So not writing them back to PTY							
							debug("Skipping RIM packet ending by "+RIM_PACKET_TAIL)
						else:
							data=array.array("B",bytes)
							os.write(self.output,data.tostring())
				except usb.USBError, error:
					# Ignore the odd "No error" error, must be a pyusb bug ?
					if error.message != "No error":
						raise
								
			except:
				if(self.done):
					# Ignoring exception during shutdown attempt
					pass
				else:
					raise
	
''' 'Static' Methods '''
def debug(msg):
	print msg
	
def debug_bytes(tuple, msg):
	'''Get a tuple of bytes and print it as lines of 16 digits (hex and ascii)'''
	text=""
	hexa=""
	cpt=0
	for t in tuple:
		#print t
		hexa+=hex(t)+" "
		if(t>=32 and t<=126):
			text+=chr(t)
		else:
			text+="."
		cpt+=1
		if(cpt%16==0 or cpt==len(tuple)):
			debug(msg+"["+hexa+"] ["+text+"]")
			text=""
			hexa=""

def end_with_tuple(tuple1,tuple2):
	'''check if tuple1 ends with tuple2'''
	if tuple1==None and tuple2==None:
		return True
	if (tuple1==None and tuple2 != None) or (tuple1!=None and tuple2==None):
		return False
	if len(tuple1) < len(tuple2):
		return False
	return is_same_tuple(tuple1[len(tuple1)-len(tuple2):len(tuple1)],tuple2)

def is_same_tuple(tuple1,tuple2):
	'''Compare 2 tuples of Bytes, return True if Same(same data)'''
	if tuple1==None and tuple2==None:
		return True
	if (tuple1==None and tuple2 != None) or (tuple1!=None and tuple2==None):
		return False
	if len(tuple1) != len(tuple2):
		return False
	# !None and same length ... compare
	length=len(tuple1)
	for i in range(length):
		if tuple1[i] != tuple2[i]:
			return False	
	return True

# MAIN
BBTether()

'''
Protocol References: (Used to figure out BBerry protocol)
	http://www.off.net/cassis/protocol-description.html
	http://xmblackberry.cvs.sourceforge.net/viewvc/xmblackberry/XmBlackBerry/	
	http://barry.cvs.sourceforge.net/viewvc/barry/barry/
	http://libusb.sourceforge.net/doc/
	http://bazaar.launchpad.net/~pygarmin-dev/pygarmin/trunk/annotate/91?file_id=garmin.py-20070323161514-arelz0uc976re3e4-1
'''
